### ISSUE TITLE
Cannot receive incremental stream: dataset is busy

### ISSUE THREAD

**Issue by ZackaryWelch** [zfsonfreebsd/ZoF#31]
Some tests in rsend fail due a 'zfs receive' command not being able to get input due to a busy dataset. These tests are rsend_001 through rsend_007, rsend_011 through rsend_013, send-c_props, send-c_incremental, and send-c_verify_contents. The full error looks like:
```
cannot receive incremental stream: dataset is busy
ERROR: eval zfs receive -d -F testpool2 < /tests/backdir-rsend/pool-final-R exited 1
```

**Comment by ZackaryWelch** [comment]
For rsend_001:
```
(in setup)
zpool create testpool /dev/md0
zpool create testpool2 /dev/md1
zpool create testpool3 /dev/md2
mkdir /var/tmp/backdir-rsend /var/tmp/tests
zfs snapshot -r testpool@final
zfs create -p testpool/testfs/fs1/fs2
zfs create -V 16M testpool/vol
zfs create -V 16M testpool/testfs/vol
zfs snapshot testpool/testfs/vol@vsnap
zfs clone testpool/testfs/vol@vsnap testpool/testfs/vclone
zfs snapshot testpool/vol@snapA
zfs snapshot testpool/testfs/vol@snapB
zfs snapshot testpool/testfs/vol@snapC
zfs snapshot testpool/testfs/vclone@snapC
zfs snapshot -r testpool@final
(in test)
zfs send -R testpool@final > /var/tmp/backdir-rsend/pool-final-R
zfs receive -d -F testpool2 < /var/tmp/backdir-rsend/pool-final-R
```

**Comment by mattmacy** [comment]
I don't see any errors. What am I supposed to see?

```
root@beastie3:/home/mmacy # zpool create -f testpool /dev/md0                                                                                       
root@beastie3:/home/mmacy # zpool create -f testpool2 /dev/md1                                                                                      
root@beastie3:/home/mmacy # zpool create -f testpool3 /dev/md2                                                                                      
root@beastie3:/home/mmacy # mkdir /var/tmp/backdir-rsend
root@beastie3:/home/mmacy # zfs snapshot -r testpool@final
root@beastie3:/home/mmacy # zfs create -p testpool/testfs/fs1/fs2
root@beastie3:/home/mmacy # zfs create -V 16M testpool/vol
root@beastie3:/home/mmacy # zfs create -V 16M testpool/testfs/vol
root@beastie3:/home/mmacy # zfs send -R testpool@final > /var/tmp/backdir-rsend/pool-final-R
root@beastie3:/home/mmacy # zfs receive -d -F testpool2 < /var/tmp/backdir-rsend/pool-final-R 
```

**Comment by ZackaryWelch** [comment]
The error would be 'Cannot receive incremental stream: dataset is busy ', which fails the 'zfs receive' command. I updated the command list with more snapshots, but they don't seem necessary for the test.

**Comment by mattmacy** [comment]
@ZackaryWelch are you able to reproduce?

**Comment by mattmacy** [comment]
@ZackaryWelch I'm closing if you can't reproduce

**Comment by mattmacy** [comment]
@kmoore134 is this still an issue?

**Comment by kmoore134** [comment]
@mattmacy Confirmed this still fails on latest:

```
% ./zfs list
NAME                         USED  AVAIL     REFER  MOUNTPOINT
testpool                    37.9M  3.59G       42K  /testpool
testpool/pclone               16K  3.59G       26K  /testpool/pclone
testpool/testfs             19.0M  3.59G       35K  /testpool/testfs
testpool/testfs/fs1          415K  3.59G      291K  /testpool/testfs/fs1
testpool/testfs/fs1/fclone    52K  3.59G       66K  /testpool/testfs/fs1/fclone
testpool/testfs/fs1/fs2       29K  3.59G       29K  /testpool/testfs/fs1/fs2
testpool/testfs/vclone         0B  3.59G       12K  -
testpool/testfs/vol         18.5M  3.61G       12K  -
testpool/vol                18.5M  3.61G       12K  -
testpool2                    142K  3.62G       24K  /testpool2
testpool2/sendfs              24K  3.62G       24K  /testpool2/sendfs

% sudo ./zfs snapshot -r testpool/testfs@snap1                                                                                                    
% sudo ./zfs send -R testpool/testfs@snap1 > /tmp/poolsnap
% sudo ./zfs receive -d -F testpool2 < /tmp/poolsnap
cannot receive incremental stream: dataset is busy
```
Note that after the failure the /testpool2 does show some of the receive did occur. It looks like it failed during the restore of the zvol "vol" or possibly the "vclone". Suggest you try to reproduce when there is a zvol in the replication stream. Guessing the FreeBSD devfs stuff to create /dev/zvol/ devices may suspect here. But just a total shot in the dark. 

```
% ./zfs list
NAME                          USED  AVAIL     REFER  MOUNTPOINT
testpool                     38.8M  3.59G       42K  /testpool
testpool/pclone                16K  3.59G       26K  /testpool/pclone
testpool/testfs              19.0M  3.59G       35K  /testpool/testfs
testpool/testfs/fs1           415K  3.59G      291K  /testpool/testfs/fs1
testpool/testfs/fs1/fclone     52K  3.59G       66K  /testpool/testfs/fs1/fclone
testpool/testfs/fs1/fs2        29K  3.59G       29K  /testpool/testfs/fs1/fs2
testpool/testfs/vclone          0B  3.59G       12K  -
testpool/testfs/vol          18.5M  3.60G       12K  -
testpool/vol                 18.5M  3.60G       12K  -
testpool2                    20.3M  3.60G       24K  /testpool2
testpool2/sendfs               24K  3.60G       24K  /testpool2/sendfs
testpool2/testfs             19.0M  3.60G       35K  /testpool2/testfs
testpool2/testfs/fs1          416K  3.60G      292K  /testpool2/testfs/fs1
testpool2/testfs/fs1/fclone    52K  3.60G       66K  /testpool2/testfs/fs1/fclone
testpool2/testfs/fs1/fs2       29K  3.60G       29K  /testpool2/testfs/fs1/fs2
testpool2/testfs/vol         18.5M  3.62G       12K  -
```

**Comment by mattmacy** [comment]
fixed by 557e84b9e9c1abc88fdc23318c2c36bf009d2732
